### DISCUSSION TITLE
What is a "follow request" in terms of activity pub?

### DISCUSSION THREAD

**Discussion by emad-elsaid** [mastodon/mastodon#23103] [Q&A] [answered]
* I'm trying to implement activityPub in a static website
* So far I have a valid webfinger endpoint that shows up in mastodon:
  * the website: https://www.emadelsaid.com/
  * the handle: [@Emad@www.emadelsaid.com](https://www.emadelsaid.com/.well-known/webfinger) 
```json
{
  "subject": "acct:Emad@www.emadelsaid.com",
  "aliases": [
    "https://www.emadelsaid.com",
    "https://www.emadelsaid.com/+/activitypub/@Emad"
  ],
  "links": [
    {
      "href": "https://www.emadelsaid.com",
      "rel": "http://webfinger.net/rel/profile-page",
      "type": "text/html"
    },
    {
      "href": "https://www.emadelsaid.com/+/activitypub/@Emad",
      "rel": "self",
      "type": "application/activity+json"
    },
    {
      "rel": "http://ostatus.org/schema/1.0/subscribe",
      "template": "https://www.emadelsaid.com/authorize_interaction?uri={uri}"
    }
  ]
}
```
  * webfinger response refer to this endpoint: https://www.emadelsaid.com/+/activitypub/@Emad/
```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "id": "https://www.emadelsaid.com/+/activitypub/@Emad",
  "type": "Person",
  "preferredUsername": "Emad",
  "name": "Emad",
  "summary": "A Software Engineer. Ruby Golang. 🇪🇬 Egyptian 📷 Youtuber 🎮 Casual gamer and 🚀 Sci-Fi addict. My opinions are my own",
  "url": "https://www.emadelsaid.com",
  "inbox": "https://www.emadelsaid.com/+/activitypub/@Emad/inbox",
  "outbox": "https://www.emadelsaid.com/+/activitypub/@Emad/outbox",
  "endpoints": {
    "sharedInbox": "https://www.emadelsaid.com/+/activitypub/@Emad/inbox"
  },
  "icon": {
    "mediaType": "image/png",
    "type": "Image",
    "url": "https://www.emadelsaid.com/public/logo.png"
  },
  "image": {
    "mediaType": "image/png",
    "type": "Image",
    "url": "https://www.emadelsaid.com/public/logo.png"
  }
}
```
  * example on toot.community https://toot.community/@Emad@www.emadelsaid.com
* The website also implements a public outbox that includes all posts : https://www.emadelsaid.com/+/activitypub/@Emad/outbox/
* here is one of the posts: 
```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "id": "https://www.emadelsaid.com/+/activitypub/@Emad/outbox/9",
  "type": "OrderedCollectionPage",
  "prev": "https://www.emadelsaid.com/+/activitypub/@Emad/outbox/8",
  "next": "https://www.emadelsaid.com/+/activitypub/@Emad/outbox/10",
  "partOf": "https://www.emadelsaid.com/+/activitypub/@Emad/outbox",
  "orderedItems": [
    {
      "id": "https://www.emadelsaid.com/Open%20Source%20Project",
      "type": "Create",
      "actor": "https://www.emadelsaid.com/+/activitypub/@Emad",
      "published": "2023-01-11T13:57:22Z",
      "to": [
        "https://www.w3.org/ns/activitystreams#Public"
      ],
      "Object": {
        "id": "https://www.emadelsaid.com/Open%20Source%20Project",
        "type": "Note",
        "published": "2023-01-11T13:57:22Z",
        "url": "https://www.emadelsaid.com/Open%20Source%20Project",
        "attributedTo": "https://www.emadelsaid.com/+/activitypub/@Emad",
        "to": [
          "https://www.w3.org/ns/activitystreams#Public"
        ],
        "Content": "Open Source Project\nhttps://www.emadelsaid.com/Open%20Source%20Project"
      }
    }
  ]
}
```

# The problem

* I noticed that the posts don't show up on the profile on the mastodon instance (the number of posts is accurate)
* New posts don't show up too. I thought maybe old posts only won't be synced from the outbox but I noticed new posts too don't show up.
* the "follow" button turned to "withdraw follow request" 

# Questions

* What's follow request in terms of activitypub and why is it needed if the inbox already includes public posts
* do I have to implement the endpoint mastodon hits when clicking "follow"? and what do you think it should respond with? 
* do I have to send the posts to every instance that sends follow request? and why the instance can't sync periodically from the public inbox like RSS used to do. 

It's more of a general confusion about this follow request in relation to public activity streams I hope I articulated my question well. and thanks in advance.

**Comment by nightpool** [comment, marked as answer]
ActivityPub is a pub/sub protocol, and it requires a dynamic server that can respond to activities and send new content by POSTing activities to your follower's inboxes. The recommendation for static websites is to partner with a dynamic "inbox provider" that can handle notifying followers about your new content while having your JSON hosted statically.

> What's follow request in terms of activitypub and why is it needed if the inbox already includes public posts

A follow request is how your followers subscribe to be notified about your new content. When a user wants to follow someone, they send a Follow activity to that actor's inbox. Then, the actor responds with an Accept { Follow } to show that it accepts the follow request. (This is used for e.g. private followers-only posts, to ensure that the user has the chance to approve follow requests if they choose to). Now, whenever the actor has a new post to deliver, it has the list of all follower's inboxes, and sends a Create activity with an object representing the new post.

> do I have to send the posts to every instance that sends follow request? and why the instance can't sync periodically from the public inbox like RSS used to do.

Because ActivityPub is a real-time pub/sub protocol. If you want a periodic syncing protocol, then there's nothing preventing you from using RSS instead, there are plenty of RSS sites still available on the internet. ActivityPub is not that.

**Reply by emad-elsaid** [reply]
Thanks for taking the time to clarify my questions. I have couple follow up questions if you don't mind:

- How and when is the GET outbox used (in general and in the case of mastodon)
- Do you have an example of the `Accept { Follow }` object?
- I already support RSS too in the same website/implementation. and wanted to cover ActivityPub too if possible.
- Do you have an example of "inbox provider"?

**Reply by puck2** [reply]
@nightpool could you clarify this:

_"The recommendation for static websites is to partner with a dynamic "inbox provider" that can handle notifying followers about your new content while having your JSON hosted statically."_

I don't understand. I'm self-hosting wordpress and having the "follow request" problem documented here.
